### PR TITLE
Authlib implementation

### PR DESCRIPTION
**Pull request by davidkirwan**
PR to aid with internal review before making a PR back upstream.

### CODE REVIEW

File: flask_oidc/__init__.py [hunk 1]
```diff
@@ -486,268 +156,18 @@ def require_login(self, view_func):
...
-
-    def require_keycloak_role(self, client, role):
-        """
-        Function to check for a KeyCloak client role in JWT access token.
-
-        This is intended to be replaced with a more generic 'require this value
-        in token or claims' system, at which point backwards compatibility will
-        be added.
-
-        .. versionadded:: 1.5.0
```
Comment (abompard) [inline flask_oidc/__init__.py:510]: This method is public and should probably be kept, as some apps may be using it (even though I don't think any Fedora apps use it, but flask-oidc is used outside Fedora too).

File: flask_oidc/__init__.py [hunk 2]
```diff
@@ -486,268 +156,18 @@ def require_login(self, view_func):
...
-        return True
-
-    WRONG_GOOGLE_APPS_DOMAIN = 'WRONG_GOOGLE_APPS_DOMAIN'
-
-    def custom_callback(self, view_func):
-        """
-        Wrapper function to use a custom callback.
-        The custom OIDC callback will get the custom state field passed in with
-        redirect_to_auth_server.
-        """
```
Comment (abompard) [inline flask_oidc/__init__.py:674]: I'm not sure what this does but let's try to keep it if we can.

File: flask_oidc/__init__.py [hunk 3]
```diff
@@ -762,165 +182,6 @@ def logout(self):
...
-        :param scopes_required: List of scopes that are required to be
-            granted by the token before returning True.
-        :type scopes_required: list
-
-        :returns: True if the token was valid and contained the required
-            scopes. An ErrStr (subclass of string for which bool() is False) if
-            an error occured.
-        :rtype: Boolean or String
-
-        .. versionadded:: 1.1
```
Comment (abompard) [inline flask_oidc/__init__.py:785]: Let's translate that to the equivalent in Authlib.

File: flask_oidc/__init__.py [hunk 4]
```diff
@@ -762,165 +182,6 @@ def logout(self):
...
-            granted by the token before being allowed to call the protected
-            function.
-        :type scopes_required: list
-        :param render_errors: Whether or not to eagerly render error objects
-            as JSON API responses. Set to False to pass the error object back
-            unmodified for later rendering.
-        :type render_errors: callback(obj) or None
-
-        .. versionadded:: 1.0
-        """
```
Comment (abompard) [inline flask_oidc/__init__.py:874]: Same here, let's translate that to the equivalent in Authlib.

File: flask_oidc/__init__.py [hunk 5]
```diff
@@ -762,165 +182,6 @@ def logout(self):
...
-                request['client_secret'] = self.client_secrets['client_secret']
-
-        resp, content = httplib2.Http().request(
-            self.client_secrets['token_introspection_uri'], 'POST',
-            urlencode(request), headers=headers)
-        # TODO: Cache this reply
-        return _json_loads(content)
+        session.pop("token", None)
+        session.pop("userinfo", None)
+        return redirect("/")
```
Comment (abompard) [inline flask_oidc/__init__.py:187]: It does not look like the previous `logout()` method redirected the user anywhere, did it?

File: tests/client_secrets.json [hunk 6]
```diff
@@ -1,12 +1,9 @@
...
-        "auth_uri": "https://test/auth",
-        "client_id": "MyClient",
-        "client_secret": "MySecret",
-        "redirect_uris": [
-            "http://localhost:5000/oauth2callback"
-        ],
-        "token_uri": "https://test/token",
-        "token_introspection_uri": "https://test/tokeninfo"
-    }
+  "OIDC_CLIENT_ID": "MyClient",
```
Comment (abompard) [inline tests/client_secrets.json:2]: This should not be changed, as this project does not control the format of `client_secrets.json`.

File: tests/requirements-tests.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+Authlib
+pytest-django
+pytest
+six
```
Comment (abompard) [inline tests/requirements-tests.txt:4]: Do we still use six?

File: tests/requirements-tests.txt [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+nose
```
Comment (abompard) [inline tests/requirements-tests.txt:5]: Do we use nose?

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,14 @@
+[tox]
+isolated_build = True
+envlist =
+    py{310}
```
Comment (abompard) [inline tox.ini:4]: Let's try to be compatible with Python 3.7+ (all the currently supported python versions)

File: README.rst [hunk 10]
```diff
@@ -20,7 +20,6 @@ This library should work with any standards compliant OpenID Connect provider.
 
 It has been tested with:
 
-* `Google+ Login <https://developers.google.com/accounts/docs/OAuth2Login>`_
```
Comment (abompard) [inline README.rst:23]: I think we should try to keep existing features if we can. Flask-OIDC is used outside Fedora.

File: example.py [hunk 11]
```diff
@@ -1,77 +0,0 @@
...
-# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
-# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
-# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
-# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
-# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
-# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
-# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
-# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-
```
Comment (abompard) [inline example.py:24]: Why remove this file? Ideally it would keep working as-is.

**Review by abompard — DISMISSED** [review]
Alright, I've quite a few comments about backwards compatibility. I think we should try to keep the existing features available if we can, for Fedora apps which may be using them but also apps outside Fedora which use flask-oidc.
Thanks!

File: README.rst (hunk 10 above)
Comment (davidkirwan) [reply README.rst:23]: We've not tested latest version with that, so best not leave it in. I think google+ is no longer even a service? I'll try put an example app together which makes use of a service like google+ and then readd at a later date though!

File: example.py (hunk 11 above)
Comment (davidkirwan) [reply example.py:24]: oh thats a typo, readding.

**Pull request merged**